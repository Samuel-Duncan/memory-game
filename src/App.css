#root {
  margin: 0 auto;
  text-align: center;
}

.loading {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}
