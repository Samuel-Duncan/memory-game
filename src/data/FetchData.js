class FetchData {
  static async fetchData(url) {
    try {
      const response = await fetch(url);
      if (!response.ok) throw new Error(`Error fetching data from ${url}`);
      const data = await response.json();
      return data;
    } catch (error) {
      console.error(error);
      throw error;
    }
  }

  static processData(results, gameIds) {
    return results
      .filter((result) => gameIds.includes(result.id))
      .map((result) => ({
        id: result.id,
        name: result.name,
        imageUrl: result.background_image,
      }));
  }

  static async fetchAndProcessData(url, gameIds) {
    const response = await this.fetchData(url);
    return this.processData(response.results, gameIds);
  }

  static async getAllData() {
    const baseUrl = `https://api.rawg.io/api/games?key=27516e003a624c4389aa6cde8779b5b6&page=`;
    const totalPages = 14;
    const gameIds = [
      3498, 3328, 28, 41494, 290856, 3387, 28568, 2093, 326243, 3497, 857,
      28199,
    ];
    let allResults = [];

    try {
      const pagePromises = [];

      // Prepare promises for fetching and processing data for each page concurrently
      for (let page = 1; page <= totalPages; page++) {
        const pageUrl = baseUrl + page;
        pagePromises.push(this.fetchAndProcessData(pageUrl, gameIds));
      }

      // Fetch and process data for all pages concurrently
      const pageResponses = await Promise.all(pagePromises);
      pageResponses.forEach((response) => {
        allResults = [...allResults, ...response];
      });

      return allResults;
    } catch (error) {
      console.error(error);
      throw error;
    }
  }

  static async getData() {
    try {
      const allData = await this.getAllData();
      return allData;
    } catch (error) {
      console.error('Error fetching and processing data:', error);
    }
  }
}

export { FetchData };
