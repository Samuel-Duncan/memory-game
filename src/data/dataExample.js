const dataExample = [
  {
    id: 3498,
    name: 'Grand Theft Auto V',
    imageUrl:
      'https://media.rawg.io/media/games/20a/20aa03a10cda45239fe22d035c0ebe64.jpg',
  },
  {
    id: 3328,
    name: 'The Witcher 3: Wild Hunt',
    imageUrl:
      'https://media.rawg.io/media/games/618/618c2031a07bbff6b4f611f10b6bcdbc.jpg',
  },
  {
    id: 28,
    name: 'Red Dead Redemption 2',
    imageUrl:
      'https://media.rawg.io/media/games/511/5118aff5091cb3efec399c808f8c598f.jpg',
  },
  {
    id: 41494,
    name: 'Cyberpunk 2077',
    imageUrl:
      'https://media.rawg.io/media/games/26d/26d4437715bee60138dab4a7c8c59c92.jpg',
  },
  {
    id: 290856,
    name: 'Apex Legends',
    imageUrl:
      'https://media.rawg.io/media/games/737/737ea5662211d2e0bbd6f5989189e4f1.jpg',
  },
  {
    id: 3387,
    name: 'Bloodborne',
    imageUrl:
      'https://media.rawg.io/media/games/214/214b29aeff13a0ae6a70fc4426e85991.jpg',
  },
  {
    id: 28568,
    name: "Assassin's Creed II",
    imageUrl:
      'https://media.rawg.io/media/games/1be/1bed7fae69d1004c09dfe1101d5a3a94.jpg',
  },
  {
    id: 2093,
    name: "No Man's Sky",
    imageUrl:
      'https://media.rawg.io/media/games/174/1743b3dd185bda4a7be349347d4064df.jpg',
  },
  {
    id: 3497,
    name: 'Far Cry 4',
    imageUrl:
      'https://media.rawg.io/media/games/b39/b396dac1f3e0f538841aa0355dd066d3.jpg',
  },
  {
    id: 326243,
    name: 'Elden Ring',
    imageUrl:
      'https://media.rawg.io/media/games/b29/b294fdd866dcdb643e7bab370a552855.jpg',
  },
  {
    id: 857,
    name: 'Halo: The Master Chief Collection',
    imageUrl:
      'https://media.rawg.io/media/games/c24/c24f4434882ae9c2c8d9d38de82cb7a5.jpg',
  },
  {
    id: 28199,
    name: 'Ori and the Will of the Wisps',
    imageUrl:
      'https://media.rawg.io/media/games/718/71891d2484a592d871e91dc826707e1c.jpg',
  },
];

export { dataExample };
