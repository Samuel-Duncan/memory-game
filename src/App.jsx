import { useState, useEffect } from 'react';
import { FetchData } from './data/FetchData';
import { Header } from './components/Header';
import { Cards } from './components/Cards';
import { Modal } from './components/Modal';
import './App.css';

const getData = () => {
  const [cards, setCards] = useState(null);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(true);

  //Fetch data
  useEffect(() => {
    FetchData.getData()
      .then((result) => {
        if (result.status >= 400) {
          throw new Error('server error');
        }
        setCards(result);
      })
      .catch((error) => {
        setError(error);
      })
      .finally(() => setLoading(false));
  }, []);

  return { cards, error, loading };
};

function App() {
  const { cards, error, loading } = getData();
  const [score, setScore] = useState(0);
  const [bestScore, setBestScore] = useState(0);
  const [clickedCards, setClickedCards] = useState([]);
  const [isGameOver, setIsGameOver] = useState(false);
  const [isGameWon, setIsGameWon] = useState(false);

  // Check if clicked cards has duplicates (game over)
  useEffect(() => {
    if (hasDuplicates(clickedCards) || clickedCards.length === 12) {
      handleGameOver();
    }
  }, [clickedCards]);

  const hasDuplicates = (array) => {
    return array.length !== new Set(array).size;
  };

  // Handle Game Over
  const handleGameOver = () => {
    setScore(0);
    setClickedCards([]);
    setIsGameOver(true);

    if (score === 12) {
      setIsGameWon(true);
    }

    if (score > bestScore) {
      setBestScore(score);
    }
  };

  //Shuffle cards
  const shuffleCards = (array) => {
    for (let i = array.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [array[i], array[j]] = [array[j], array[i]];
    }
    return array;
  };

  const handleCardClick = (cardId) => {
    if (hasDuplicates([...clickedCards, cardId]) || score === 12) {
      handleGameOver();
    } else {
      setScore(score + 1);
      setClickedCards((prevClickedCards) => [...prevClickedCards, cardId]);
      setCards(shuffleCards([...cards]));
    }
  };

  const handlePlayAgainClick = () => {
    setIsGameOver(false);
    setIsGameWon(false);
  };

  if (error) return <p>A network error was encountered</p>;
  if (loading)
    return (
      <>
        <Header score={score} bestScore={bestScore} />
        <h2 className="loading">Loading...</h2>
      </>
    );

  return (
    <>
      <Header score={score} bestScore={bestScore} />
      <Cards data={cards} onClick={handleCardClick}></Cards>
      {(isGameOver || isGameWon) && (
        <Modal onClick={handlePlayAgainClick} isGameWon={isGameWon} />
      )}
    </>
  );
}

export default App;
