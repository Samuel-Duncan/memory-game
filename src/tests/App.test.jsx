import { render, screen, waitFor } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { FetchData } from '../data/FetchData';
import App from '../App';

test('modal displays correct message depending on game outcome', async () => {
  // Mock the data that would be fetched asynchronously
  const mockedGameData = [
    {
      id: 3498,
      name: 'Grand Theft Auto V',
      imageUrl:
        'https://media.rawg.io/media/games/20a/20aa03a10cda45239fe22d035c0ebe64.jpg',
    },
    {
      id: 3328,
      name: 'The Witcher 3: Wild Hunt',
      imageUrl:
        'https://media.rawg.io/media/games/618/618c2031a07bbff6b4f611f10b6bcdbc.jpg',
    },
    {
      id: 28,
      name: 'Red Dead Redemption 2',
      imageUrl:
        'https://media.rawg.io/media/games/511/5118aff5091cb3efec399c808f8c598f.jpg',
    },
    {
      id: 41494,
      name: 'Cyberpunk 2077',
      imageUrl:
        'https://media.rawg.io/media/games/26d/26d4437715bee60138dab4a7c8c59c92.jpg',
    },
    {
      id: 290856,
      name: 'Apex Legends',
      imageUrl:
        'https://media.rawg.io/media/games/737/737ea5662211d2e0bbd6f5989189e4f1.jpg',
    },
    {
      id: 3387,
      name: 'Bloodborne',
      imageUrl:
        'https://media.rawg.io/media/games/214/214b29aeff13a0ae6a70fc4426e85991.jpg',
    },
    {
      id: 28568,
      name: "Assassin's Creed II",
      imageUrl:
        'https://media.rawg.io/media/games/1be/1bed7fae69d1004c09dfe1101d5a3a94.jpg',
    },
    {
      id: 2093,
      name: "No Man's Sky",
      imageUrl:
        'https://media.rawg.io/media/games/174/1743b3dd185bda4a7be349347d4064df.jpg',
    },
    {
      id: 3497,
      name: 'Far Cry 4',
      imageUrl:
        'https://media.rawg.io/media/games/b39/b396dac1f3e0f538841aa0355dd066d3.jpg',
    },
    {
      id: 326243,
      name: 'Elden Ring',
      imageUrl:
        'https://media.rawg.io/media/games/b29/b294fdd866dcdb643e7bab370a552855.jpg',
    },
    {
      id: 857,
      name: 'Halo: The Master Chief Collection',
      imageUrl:
        'https://media.rawg.io/media/games/c24/c24f4434882ae9c2c8d9d38de82cb7a5.jpg',
    },
    {
      id: 28199,
      name: 'Ori and the Will of the Wisps',
      imageUrl:
        'https://media.rawg.io/media/games/718/71891d2484a592d871e91dc826707e1c.jpg',
    },
  ];

  const originalGetData = FetchData.getData;

  // Mock the FetchData.getData() method to return a resolved Promise with the mocked game data
  FetchData.getData = () => Promise.resolve(mockedGameData);

  // Render the App component
  render(<App />);

  // Wait for the data to be rendered
  await waitFor(() => screen.getByText('Grand Theft Auto V'));

  // Simulate playing the game
  for (const item of mockedGameData) {
    userEvent.click(screen.getByText(item.name));
  }

  // Wait for modal to appear
  await waitFor(() => screen.getByText('You won!'));

  // Ensure modal displays the correct message for game won
  expect(screen.getByText('You won!')).toBeInTheDocument();

  // Restore the original implementation of FetchData.getData
  FetchData.getData = originalGetData;
});
