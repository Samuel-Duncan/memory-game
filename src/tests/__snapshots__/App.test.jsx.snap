// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > matches snapshot 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="header-left"
    >
      <h1>
        RPG Memory Game
      </h1>
    </div>
    <div
      class="header-right"
    >
      <span>
        Score: 
        0
      </span>
      <span>
        Best score: 
        0
      </span>
    </div>
  </header>
  <div
    class="cards"
  />
</div>
`;
